server/src/main/java/org/elasticsearch/search/runtime/AbstractStringScriptFieldAutomatonQuery.java

```diff
@@ -17,7 +17,7 @@
 import java.util.List;
 
 public abstract class AbstractStringScriptFieldAutomatonQuery extends AbstractStringScriptFieldQuery {
-    private final BytesRefBuilder scratch = new BytesRefBuilder();
+    private final ThreadLocal<BytesRefBuilder> bytesRefBuilderThreadLocal = ThreadLocal.withInitial(BytesRefBuilder::new);
```

Comment:
@nik9000 pinging you because you wrote this in the first place. How important is it to have a pre-initialized BytesRefBuilder as opposed to creating one within matches every time? I went for the ThreadLocal initially, but I think that it's probably overkill. Do you have opinions?

Reply:
Thinking out loud: before we would create a single scratch, shared across all segments. With the thread local, we would create one per thread executing a certain query: there's a max number of threads we use for a single search, despite the number of slices may be high. Without the threadlocal, we would create a new instance for each segment.

Reply:
I think it's pretty important not to allocate every time. I just fixed someone on the ESQL side that is shaped kind of similar and it went from ~25 nanos per run to ~12 nanos per run. 

If we have segment-level concurrency we could make the scratch when we make a `Scorer`.

Reply:
> If we have segment-level concurrency we could make the scratch when we make a Scorer.

How would that work? The scorer is made per segment anyways?  Each thread that runs matches needs to have a different instance of the scratch, it cannot be shared, unless we want to synchronize its access, which does not seem like a good idea. That is what the thread local does: give each thread a different instance.

Reply:
Yeah, Scorers are always made per segment. If we're running per-segment concurrency that's probably fine.

Reply:
> One scratch per segment or one per thread does not make a huge difference.

Yeah! I really like avoiding the ThreadLocal because those are hard for me to read.

Reply:
yea I tend to agree. I extended the script context, that gets created per segment, to include the scratch instance. This should do it.

Reply:
err, that does not work, and reminds me once again never to mess with painless script contexts :)